subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java

```diff
@@ -311,7 +318,29 @@ public void visitInputFileProperties(InputFilePropertyVisitor visitor) {
         @Override
         public void visitOutputProperties(OutputPropertyVisitor visitor) {
             for (OutputFilePropertySpec property : context.getTaskProperties().getOutputFileProperties()) {
-                visitor.visitOutputProperty(property.getPropertyName(), property.getOutputType(), property.getPropertyFiles());
+                ImmutableSet.Builder<File> builder = ImmutableSet.builder();
+                ((FileCollectionInternal) property.getPropertyFiles()).visitStructure(new FileCollectionStructureVisitor() {
+                    @Override
+                    public void visitCollection(FileCollectionInternal.Source source, Iterable<File> contents) {
+                        builder.addAll(contents);
+                    }
+
+                    @Override
+                    public void visitGenericFileTree(FileTreeInternal fileTree) {
+                        builder.addAll(fileTree);
+                    }
+
+                    @Override
+                    public void visitFileTree(File root, PatternSet patterns, FileTreeInternal fileTree) {
+                        builder.add(root);
+                    }
+
+                    @Override
+                    public void visitFileTreeBackedByFile(File file, FileTreeInternal fileTree) {
+                        builder.add(file);
+                    }
+                });
+                visitor.visitOutputProperty(property.getPropertyName(), property.getOutputType(), builder.build());
```

Comment:
Should we do this fix on `master` separately?

Reply:
PR: https://github.com/gradle/gradle/pull/10901

Reply:
The PR has been merged.